### PR TITLE
Add `using_terminal` stanza

### PR DESCRIPTION
**Pull request by CT075**
As per #3464 .

Todo:
- [x] Capture standard terminal output while the terminal lock is held
- [ ] Add tests

### CODE REVIEW

File: src/dune/action_exec.ml [hunk 1]
```diff
@@ -3,6 +3,9 @@ open Import
 open Fiber.O
 module DAP = Dune_action_plugin.Private.Protocol
 
+(* This feels like the wrong place for this, but I'm not sure of a better one. *)
+let terminal_lock = Fiber.Mutex.create ()
```
Comment (jeremiedimino) [inline src/dune/action_exec.ml:7]: Looks about right to me. The terminal is a global resource, so it seems reasonable to use a global lock.

Comment (jeremiedimino) [reply src/dune/action_exec.ml:7]: However, yes this might not be the right module to put it in.

**Comment by jeremiedimino** [comment]
That looks right to me, but we also need to prevent dune from printing messages while such an action is running. This should be done globally via the `Console` module in `src/stdune/console.mli`.

**Comment by CT075** [comment]
> That looks right to me, but we also need to prevent dune from printing messages while such an action is running. This should be done globally via the `Console` module in `src/stdune/console.mli`.

I'm unsure of what the behavior in this case should be. Should we silently eat anything sent to the console during such an action, or capture them to be output later?

**Comment by rgrinberg** [comment]
> Should we silently eat anything sent to the console during such an action, or capture them to be output later?

The latter seems much more reasonable to me.

**Comment by rgrinberg** [comment]
Quoting Jeremie from the original issue:

> Only one such action can run at a time, and while such an action is running messages emitted by dune would be put in a queue and rendered once the action has finished.

**Comment by rgrinberg** [comment]
> Adjust dune/context.ml to possibly take into account the selective terminal usage?

What needs to be adjust in `context.ml`?

**Comment by CT075** [comment]
> > Adjust dune/context.ml to possibly take into account the selective terminal usage?
> 
> What needs to be adjust in `context.ml`?

This was an idea I had when I was far less familiar with how the codebase functioned; I'm no longer sure what I really meant. I'll remove it from the PR description.

**Comment by CT075** [comment]
I'd really like to make `Console.lock` and `Console.unlock` idempotent (or, even better, expose only a safe interface for buffering, but I suspect that this would require pulling in `Fiber`), but I'm not really sure how a buffered console should interact with console composition.

**Comment by CT075** [comment]
Wait, for that matter, what is terminal composition currently useful for? Neither grepping over the codebase or using github search seems to come up with any uses of it.

File: src/dune_engine/action_exec.ml [hunk 2]
```diff
@@ -355,6 +357,19 @@ let rec exec t ~ectx ~eenv =
...
+        Console.lock ();
+        let* result =
+          exec t ~ectx
+            ~eenv:
+              { eenv with
+                stdout_to = Process.Io.stdout
+              ; stderr_to = Process.Io.stderr
+              }
+        in
+        Console.unlock ();
```
Comment (rgrinberg) [inline src/dune_engine/action_exec.ml:371]: What happens when the `exec` above raises? Shouldn't we still unlock?

File: src/stdune/console.mli [hunk 3]
```diff
@@ -41,6 +41,17 @@ include Backend.S
     {[ print_user_message (User_message.make paragraphs) ]} *)
 val print : User_message.Style.t Pp.t list -> unit
 
+(** [lock] captures any messages written to the console, which are rendered in
+    the previous console mode upon being [unlock]ed.
+
+    To allow these to be used alongside [Fiber] jobs, we expose this unsafe
+    interface instead of [with_lock: (unit -> unit) -> unit]. Double-[lock]ing
+    or [unlock]ing the console will lead to output errors (no output, repeated
+    output, etc). *)
```
Comment (rgrinberg) [inline src/stdune/console.mli:50]: Can we just forbid double locking/unlocking via runtime checks?

Comment (cwong-ocaml) [reply src/stdune/console.mli:50]: This was also my idea, but I wasn't sure what to do in the case of composing a buffered terminal with a non-buffered terminal.

File: src/dune_engine/action_exec.ml (hunk 2 above)
Comment (cwong-ocaml) [reply src/dune_engine/action_exec.ml:371]: Good catch.

Comment (CT075) [reply src/dune_engine/action_exec.ml]: Good catch, fixed.

File: src/stdune/console.mli (hunk 3 above)
Comment (CT075) [reply src/stdune/console.mli]: I spoke to @jeremiedimino about the behavior of terminal locking in the presence of composed terminals, and we concluded that we don't need to worry about it for now, so I've adjusted the interface.

**Comment by rgrinberg** [comment]
Please re-open if this is still relevant

**Pull request closed without merging**